### ISSUE TITLE
Its hard to extend components due to many private methods / properties

### ISSUE THREAD

**Issue by brstgt** [amphp/aerys#175]
Examples:

- Copy watchers from an existing loop to new loop
- Extend WatcherProcess to modify the worker script to use a custom WorkerProcess

There are probably more examples but these are on the top of my head

**Comment by kelunik** [comment]
Why do you want to replace an existing event loop and copy the watchers?

Which extensions are you looking for in `WatcherProcess`?

**Comment by brstgt** [comment]
1. I tried to replace an existing loop while the loop was already running but the existing watchers were then dead and I had no way to "copy" them over to the new loop. I ran into that issue yesterday. Currently the only way to fix it was to replace the loop before any watchers were added. The custom loop is "only" a decorator but as a driver is not only an interface but MUST extend a abstract implementation with its own state + private methods, it's not possible to FULLY decorate the old driver. All in all it's not really clean. Maybe the cleanest approach was to have a Driver interface.
2. I wanted to implement more logic in the watcher / worker. E.g. to collect stats like req/s, busy time, kill hanging workers if they don't ping for more than $x seconds. To do that I had to implement a separate watcher process as from within the config|server I cannot reach the worker process AND the IPC protocol is quite dump at the moment
